### ISSUE TITLE
java.util.NoSuchElementException: Pool exhausted #2

### ISSUE THREAD

**Issue by mafrankowski** [messaginghub/pooled-jms#29]
I've got following exception (same as reported in https://github.com/messaginghub/pooled-jms/issues/15) from the latest pooled-jms code (3.1.0) as well as from version 1.2.4:
java.util.NoSuchElementException: Pool exhausted
at org.apache.commons.pool2.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:457)
at org.apache.commons.pool2.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:350)
at org.messaginghub.pooled.jms.JmsPoolConnectionFactory.createJmsPoolConnection(JmsPoolConnectionFactory.java:716)

or from version 1.2.4
at org.messaginghub.pooled.jms.JmsPoolConnectionFactory.createJmsPoolConnection(JmsPoolConnectionFactory.java:732)

Please use the same test method to reproduce the issue as reported in previous related issue - it should be added to org.messaginghub.pooled.jms.JmsPoolConnectionFactoryTest, optionally changing 1ms to 2ms as below because in case of 1ms it sometimes fails somewhere else. This is random issue so sometimes this tests fails sometimes it passes.

```
@Test
public void testCreateConnectionWithIdleTimeoutAndCheckInterval() throws Exception {
    cf.setConnectionIdleTimeout(2);
    cf.setConnectionCheckInterval(2);
    for (int i = 0;i< 10000;++i){
        Connection connection = cf.createConnection();
        assertNotNull(connection);
        connection.close();
        Thread.sleep(2);
    }
}
```

**Comment by tabish121** [comment]
I'm unable to reproduce this using the test provided.

**Comment by mafrankowski** [comment]
Please find a little bit changed test attached, hope you will be able to reproduce it with this one.
When I run it locally I get at least few failures out of 10 retries defined (checked it now on version 3.1.2-SNAPSHOT). Additionally it looks that changing logging to org.slf4j.simpleLogger.log.org.messaginghub.pooled.jms=TRACE even increase probability of failure.
This issue also occurs on production around once a month, of course with a little bit more appropriate eviction parameters.

[JmsPoolConnectionNoSuchElementExceptionTest.zip](https://github.com/messaginghub/pooled-jms/files/12523100/JmsPoolConnectionNoSuchElementExceptionTest.zip)

**Comment by mafrankowski** [comment]
I put idle timeout as 1 in test only to increase chances to reproduce it.
I observe the same issue around once every month on production (which is very rare but still) for following configuration

`pooledConnectionFactory.setConnectionIdleTimeout(200_000);
 pooledConnectionFactory.setConnectionCheckInterval(200_000);`

**Comment by mafrankowski** [comment]
It seems that adding Thread.yield() line as below somehow fixes the issue from test, but it is just a guess as I do not understand that logic with 10 retries when NoSuchElementException is thrown

```java
                        if (exhaustedPoolRecoveryAttempts++ < exhaustionRecoverAttemptLimit) {
                            LOG.trace("Recover attempt {} from exhausted pool by refilling pool key and creating new Connection", exhaustedPoolRecoveryAttempts);
                            connectionsPool.addObject(key);
                            Thread.yield();
                            continue;
                        } else {
```

**Comment by tabish121** [comment]
I've released version 3.1.2 with some improvements in this area to attempt to resolve this however in a worst case scenario it can still throw ResourceAllocationException errors if it cannot obtain a factory without blocking for to long.
